### PR TITLE
Development

### PR DESCRIPTION
**Pull request by Incarnation-p-lee**
* Add annotations for Graph, Vertex and Edge
* Only implement insert and deleteAll query.

### CODE REVIEW

File: config/findbugs-exclude.xml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<FindBugsFilter>
+    <Class name="com.microsoft.azure.spring.data.documentdb.common.GetHashMac"/>
```
Comment (sophiaso) [inline config/findbugs-exclude.xml:4]: Please change the class name.

**Comment by yaweiw** [comment]
@Incarnation-p-lee too much. split it into smaller PRs. normally 2-3 files each.

File: pom.xml [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    <description>Gremlin support for Spring Data</description>
+    <url></url>
+
+    <properties>
+        <maven.compiler.source>1.8</maven.compiler.source>
+        <maven.compiler.target>1.8</maven.compiler.target>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <spring.springframework.version>5.0.1.RELEASE</spring.springframework.version>
+        <spring.data.version>2.0.1.RELEASE</spring.data.version>
+        <spring.boot.starter.test.version>1.5.8.RELEASE</spring.boot.starter.test.version>
```
Comment (sophiaso) [inline pom.xml:21]: Can we use 2.0.0 for the test starter?

**Review by yaweiw — CHANGES_REQUESTED** [review]
split it into smaller PRs. normally 2-3 files each.

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>3.5</version>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot</artifactId>
+            <version>2.0.0.RELEASE</version>
```
Comment (sophiaso) [inline pom.xml:96]: Use property for version?

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,205 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>com.microsoft.spring.data.gremlin</groupId>
+    <artifactId>spring-data-gremlin</artifactId>
+    <version>1.0-SNAPSHOT</version>
```
Comment (yungezz) [inline pom.xml:9]: keep consistent with spring-data-docdb? 0.1.x and 2.0.x

File: pom.xml (hunk 2 above)
Comment (yungezz) [reply pom.xml:21]: spring data version keeps consistent with spring boot starter

File: src/main/java/com/microsoft/spring/data/gremlin/conversion/GremlinScriptGraphTraversal.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+ */
+package com.microsoft.spring.data.gremlin.conversion;
+
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversal;
+
+public class GremlinScriptGraphTraversal implements GremlinScript<GraphTraversal> {
+
+    @Override
+    public GraphTraversal generateScript(GremlinSource gremlinSource) {
+        return null;
```
Comment (sophiaso) [inline src/main/java/com/microsoft/spring/data/gremlin/conversion/GremlinScriptGraphTraversal.java:14]: Can we throw notsupported/implemented exception here?

File: src/main/java/com/microsoft/spring/data/gremlin/annotation/Edge.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE in the project root for
+ * license information.
+ */
+package com.microsoft.spring.data.gremlin.annotation;
```
Comment (yungezz) [inline src/main/java/com/microsoft/spring/data/gremlin/annotation/Edge.java:6]: so package is com.microsoft.spring.xxx or com.microsoft.azure.spring.xxx? seems first one more general

Comment (yungezz) [reply src/main/java/com/microsoft/spring/data/gremlin/annotation/Edge.java:6]: if use first one, pls update pom file to be consistent

File: src/main/java/com/microsoft/spring/data/gremlin/common/GremlinFactory.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import org.apache.tinkerpop.gremlin.driver.Client;
+import org.apache.tinkerpop.gremlin.driver.Cluster;
+import org.apache.tinkerpop.gremlin.driver.ser.Serializers;
+import org.springframework.lang.NonNull;
+
+public class GremlinFactory {
+    private Client client;
+    private Cluster gremlinCluster;
+
+    public GremlinFactory(@NonNull String endpoint, @NonNull String port,
```
Comment (yungezz) [inline src/main/java/com/microsoft/spring/data/gremlin/common/GremlinFactory.java:17]: has port default value?

File: src/main/java/com/microsoft/spring/data/gremlin/common/GremlinFactory.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+                          @NonNull String username, @NonNull String password) {
+        this.gremlinCluster = Cluster.build(endpoint).serializer(Serializers.DEFAULT_RESULT_SERIALIZER)
+                .credentials(username, password).enableSsl(true).port(Integer.parseInt(port)).create();
+
+        this.client = this.gremlinCluster.connect();
```
Comment (yungezz) [inline src/main/java/com/microsoft/spring/data/gremlin/common/GremlinFactory.java:22]: any exception handling? especially when connect, such as invalid connection string

File: src/main/java/com/microsoft/spring/data/gremlin/conversion/GremlinSourceBase.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.Map;
+
+@Getter
+@Setter
+@AllArgsConstructor
+public class GremlinSourceBase implements GremlinSource {
+
+    private String id;
+    private String label;
+    private Map properties;
```
Comment (sophiaso) [inline src/main/java/com/microsoft/spring/data/gremlin/conversion/GremlinSourceBase.java:22]: The expected behavior of this properties is unclear from the type.

File: src/main/java/com/microsoft/spring/data/gremlin/conversion/GremlinSourceBase.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public boolean hasProperty(String key) {
+        return this.properties.get(key) != null;
+    }
+
+    @Override
+    public void setProperty(String key, String value) {
+        if (this.hasProperty(key) && value != null) {
+            return;
```
Comment (sophiaso) [inline src/main/java/com/microsoft/spring/data/gremlin/conversion/GremlinSourceBase.java:37]: So the value of certain key cannot be changed once set?

File: src/main/java/com/microsoft/spring/data/gremlin/conversion/GremlinSourceBase.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+        this.properties = new HashMap<String, String>();
+    }
+
+    public boolean hasProperty(String key) {
+        return this.properties.get(key) != null;
+    }
+
+    @Override
+    public void setProperty(String key, String value) {
+        if (this.hasProperty(key) && value != null) {
```
Comment (yungezz) [inline src/main/java/com/microsoft/spring/data/gremlin/conversion/GremlinSourceBase.java:36]: property cannot be overwrite?

File: src/main/java/com/microsoft/spring/data/gremlin/conversion/GremlinSourceEdge.java [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+@AllArgsConstructor
+public class GremlinSourceEdge extends GremlinSourceBase {
+
+    private String vertexIdFrom;
+    private String vertexIdTo;
+
+    public GremlinSourceEdge() {
+        super();
+
+        this.vertexIdFrom = null;
```
Comment (sophiaso) [inline src/main/java/com/microsoft/spring/data/gremlin/conversion/GremlinSourceEdge.java:21]: By default, String field is already null?

**Comment by Incarnation-p-lee** [comment]
Just for design review, will split it when the design is stable.

File: src/main/java/com/microsoft/spring/data/gremlin/query/GremlinQueryCreator.java [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+                if (path.getType() != String.class) {
+                    throw new IllegalArgumentException("part must be String, but: " + path.getType() + ", " + path);
+                }
+
+                return criteria;
+            case WHEN_POSSIBLE:
+                return criteria;
+            case NEVER:
+                break;
+            default:
```
Comment (sophiaso) [inline src/main/java/com/microsoft/spring/data/gremlin/query/GremlinQueryCreator.java:59]: what's the default behavior?

**Pull request closed without merging**